### ISSUE TITLE
ConstructingParser throws NoSuchMethodError in Scala 3

### ISSUE THREAD

**Issue by cayhorstmann** [scala/scala-xml#541]
Save https://horstmann.com/unblog/2020-11-27/index.html to a local file and run

```scala
import java.io.File
import scala.xml.parsing.ConstructingParser
val filename="..."
val file = new File(filename)
val parser = ConstructingParser.fromFile(file, preserveWS = true)
val doc = parser.document().docElem
```

In Scala 2, this works perfectly. In Scala 3.0.0, an exception is thrown:

```
java.lang.NoSuchMethodError: 'scala.xml.parsing.MarkupParser scala.xml.parsing.MarkupParser$WithLookAhead.scala$xml$parsing$MarkupParser$WithLookAhead$$$outer()'
  at scala.xml.parsing.MarkupParser.lookahead(MarkupParser.scala:77)
  at scala.xml.parsing.MarkupParser.lookahead$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.lookahead(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParserCommon.peek(MarkupParserCommon.scala:267)
  at scala.xml.parsing.MarkupParserCommon.xTakeUntil(MarkupParserCommon.scala:249)
  at scala.xml.parsing.MarkupParserCommon.xTakeUntil$(MarkupParserCommon.scala:25)
  at scala.xml.parsing.ConstructingParser.xTakeUntil(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.xCharData(MarkupParser.scala:382)
  at scala.xml.parsing.MarkupParser.xCharData$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.xCharData(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.content1(MarkupParser.scala:427)
  at scala.xml.parsing.MarkupParser.content1$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.content1(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.content(MarkupParser.scala:462)
  at scala.xml.parsing.MarkupParser.content$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.content(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.element1(MarkupParser.scala:591)
  at scala.xml.parsing.MarkupParser.element1$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.element1(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.content1(MarkupParser.scala:436)
  at scala.xml.parsing.MarkupParser.content1$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.content1(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.content(MarkupParser.scala:462)
  at scala.xml.parsing.MarkupParser.content$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.content(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.element1(MarkupParser.scala:591)
  at scala.xml.parsing.MarkupParser.element1$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.element1(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.content1(MarkupParser.scala:436)
  at scala.xml.parsing.MarkupParser.content1$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.content1(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.content(MarkupParser.scala:462)
  at scala.xml.parsing.MarkupParser.content$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.content(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.element1(MarkupParser.scala:591)
  at scala.xml.parsing.MarkupParser.element1$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.element1(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.content1(MarkupParser.scala:436)
  at scala.xml.parsing.MarkupParser.content1$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.content1(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.content(MarkupParser.scala:462)
  at scala.xml.parsing.MarkupParser.content$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.content(ConstructingParser.scala:52)
  at scala.xml.parsing.MarkupParser.document(MarkupParser.scala:251)
  at scala.xml.parsing.MarkupParser.document$(MarkupParser.scala:33)
  at scala.xml.parsing.ConstructingParser.document(ConstructingParser.scala:52)
```

The offending code is

```
def lookahead(): BufferedIterator[Char] = curInput match {
    case curInputWLA: WithLookAhead =>
      curInputWLA.lookahead()
```

The library dependency in `build.sbt`:

```
libraryDependencies ++= Seq(
  "org.scala-lang.modules" %% "scala-xml" % "2.0.0",
)
```

**Comment by SethTisue** [comment]
I'm able to reproduce the problem locally in the Scala 3 REPL.

Once we dig into this it seems like it has to be a Scala 3 bug, but we need a scala-xml ticket anyway because we'll need to publish a new version of the library once there's a fix.

**Comment by joroKr21** [comment]
Next step would be to `getDeclaredMethods` of the `WithLookAhead` class and see what is the name of the outer accessor (if it exists)

**Comment by SethTisue** [comment]
minimized (not sure the string is absolutely minimal, but it doesn't matter, we just want something short that triggers it)

```
  @Test
  def issue541: Unit = {
    val xml =
      """|<script>// <![CDATA[
         |[]; // ]]>
         |</script>""".stripMargin
    val parser = ConstructingParser.fromSource(Source.fromString(xml), preserveWS = true)
    parser.document().docElem  // shouldn't crash
  }
```

**Comment by SethTisue** [comment]
I've reported the Scala 3 compiler bug responsible over at https://github.com/lampepfl/dotty/issues/13096

Here in this repo, I will submit a pull request with a workaround.

**Comment by SethTisue** [comment]
PR: #542

**Comment by SethTisue** [comment]
release ticket: https://github.com/scala/scala-xml/issues/543

**Comment by SethTisue** [comment]
fixed in 2.0.1 which is now published
